packages/core-transaction-pool/lib/guard.js

```diff
@@ -266,18 +266,19 @@ module.exports = class TransactionGuard {
     // Add transactions to the transaction pool
     const { added, notAdded } = this.pool.addTransactions(this.accept)
 
-    // Filter not accepted tx
-    this.accept = this.accept.filter(accepted => added.includes(accepted))
-
-    // Only broadcast accepted transactions
-    this.broadcast = this.broadcast.filter(broadcast =>
-      this.accept.includes(broadcast),
-    )
+    const removeTransaction = (tx, transactions) => {
+      const index = transactions.indexOf(tx)
+      if (index !== -1) {
+        transactions.splice(index, 1)
+      }
+    }
 
-    // Add errors
-    notAdded.forEach(error =>
-      this.__pushError(error.transaction, error.type, error.message),
-    )
+    // Exclude transactions which were refused from the pool
+    notAdded.forEach(item => {
+      removeTransaction(item.transaction, this.accept)
+      removeTransaction(item.transaction, this.broadcast)
```

Comment:
This code has complexity `notAdded.length * (accept.length + broadcast.length)`. For each element in `notAdded` we search linearly in `accept` and also search linearly in `broadcast`.

The complexity can be lowered to just `notAdded.length` if the search and removal is made constant by not using an array and using a `Set` or a plain Object {} with keys = transaction id.